### PR TITLE
[DSv4][P5-1] mxfp8_act_quant: CUDA + Triton kernels

### PR DESCRIPTION
**Pull request by ryankert01**
## Summary

P5-1: `mxfp8_act_quant_fwd` / `mxfp8_act_quant_bwd` — BF16 routed-expert activations → MXFP8 (OCP E4M3 elements, block-32 E8M0 shared scales), plus the STE backward, in **both CUDA and Triton**. Both backends are **byte-exact** with the start-kit oracle (`rl_engine.moe.mx_format.mx_quantize`) and therefore keep the `oracle-fp32-serial-v1` numeric profile — no relaxed profile is registered.

- `csrc/cuda/moe/mxfp8_act_quant.cu` — one thread owns one MX block (bf16/fp16: 4 × 16 B loads, no cross-lane reduction; fp32: 8 elements/thread + xor-shuffle on integer bit patterns)
- `rl_engine/kernels/ops/triton/moe/mxfp8_act_quant.py` — `[32, 32]` tile per program, `num_warps` switches on input size
- `rl_engine/moe/backends/mxfp8_act_quant.py` — `CudaMXFP8ActQuantProvider` / `TritonMXFP8ActQuantProvider`
- `docs/operators/mxfp8-act-quant.md`, `_C.pyi` stubs, `setup.py` / `ops.cpp` wiring

## Frozen recipe, and how the kernels stay exact

`amax` is an integer max over `|x|` bit patterns; `floor(log2 amax)` is read off the FP32 exponent field after the `FLT_MIN` clamp (the exact integer equivalent of the oracle's `frexp`); the scale is assembled from bits so code 0 is the subnormal `2**-127`; the division is an explicit non-`.ftz` `div.rn.f32` / `tl.fdiv(ieee_rounding=True)`; E4M3 is clamp-to-±448 then RNE (start-kit D1 = hardware `cvt.rn.satfinite.e4m3`). Nothing on the subnormal-sensitive path can be flushed by `--use_fast_math` (`KERNEL_ALIGN_USE_FAST_MATH=1`) — a fast-math build is a regression test.

Contract points from the P5-1 spec checked explicitly: `K % 32 == 0`; row-local amax (changing M never changes a row); non-finite input raises (fail-closed, always on in the providers); STE backward accepts any floating dtype; `act_quant1.*` / `act_quant2.*` hashed and `notes["act_quant_bwd"] == "ste"` in the trace.

## Acceptance (H100 80GB HBM3, torch 2.14+cu130, triton 3.8)

```
$ python scripts/check_p5.py --device cuda --provider rl_engine.moe.backends.mxfp8_act_quant:CudaMXFP8ActQuantProvider
provider=mxfp8-act-quant-cuda profile=oracle-fp32-serial-v1 device=cuda
RESULT: PASS (all boundaries byte-equal)
$ python scripts/check_p5.py --device cuda --provider rl_engine.moe.backends.mxfp8_act_quant:TritonMXFP8ActQuantProvider
provider=mxfp8-act-quant-triton profile=oracle-fp32-serial-v1 device=cuda
RESULT: PASS (all boundaries byte-equal)
$ pytest tests/test_p5_mxfp8_act_quant.py -q
62 passed
```

## Bit-wise alignment tests (`tests/test_p5_mxfp8_act_quant.py`)

All comparisons are `torch.equal` on raw `uint8` codes and scale bytes — never `allclose`.

| check | pins |
|---|---|
| random shapes / dtypes vs `mx_quantize` | 1×32 … 512×4096, bf16/fp16/fp32, 3-D |
| `act_quant_edges` → committed manifest | powers of two, RNE ties (17→16, 19→20), zero rows, subnormal amax hit the CPU-anchored SHA-256s |
| special values | saturating (448/464/500 → `0x7E`), wide range (2^±60), subnormal scale, all-negative |
| batch invariance | row alone == inside 257 rows == strided view |
| CUDA == Triton | the two backends cannot drift from each other |
| two outside implementations | torchao `to_mx` (FLOOR) and Triton's `triton_kernels.downcast_to_mxfp` (ROUND_DOWN) are byte-identical (skipped if not installed) |
| `--use_fast_math` build | FTZ regression |
| full routed pipeline | every boundary hash equal to the oracle's, STE noted |

## Benchmark: torch-native vs Triton vs CUDA (`benchmarks/benchmark_mxfp8_act_quant.py`)

bf16, ms/call, `check_finite=False` (the fail-closed read-back is one device sync; providers always pay it — second table):

| implementation | format | 1×7168 | 128×7168 | 4096×7168 | 16384×7168 | 8192×2048 |
|---|---|---|---|---|---|---|
| torch-native (P5 oracle) | MX | 0.339 | 0.386 | 1.361 | 4.492 | 0.906 |
| torchao `to_mx`, eager | MX | 0.155 | 0.156 | 0.681 | 2.565 | 0.413 |
| torchao `to_mx` + `torch.compile` | MX | 0.046 | 0.045 | 0.044 | 0.122 | 0.046 |
| `triton_kernels.downcast_to_mxfp` (vLLM MXFP4 path) | MX | 0.068 | 0.068 | 0.068 | 0.129 | 0.068 |
| vLLM `per_token_group_quant_fp8` (CUDA) | fp8 g128 + fp32 scale | 0.016 | 0.016 | 0.063 | 0.244 | 0.037 |
| **ours (triton)** | MX | 0.029 | 0.029 | **0.033** | **0.119** | 0.029 |
| **ours (cuda)** | MX | **0.011** | **0.011** | 0.034 | 0.119 | **0.021** |
| roofline (plain `x.to(float8_e4m3fn)`) | lower bound | 0.007 | 0.007 | 0.032 | 0.117 | 0.019 |

| with `check_finite=True` | 1×7168 | 128×7168 | 4096×7168 | 16384×7168 | 8192×2048 |
|---|---|---|---|---|---|
| ours (triton) | 0.046 | 0.046 | 0.079 | 0.165 | 0.067 |
| ours (cuda) | 0.026 | 0.026 | 0.059 | 0.146 | 0.048 |

2.99 TB/s effective at 16384×7168 — the plain-cast floor. torchao's fused MX kernels and TE's MXFP8 quantizer are gated on `sm_100`, so compiled `to_mx` and `triton_kernels` are the strongest MX baselines on Hopper; vLLM's kernel is a different format (DeepSeek-V3 recipe) and included as a cost reference.

### CODE REVIEW

File: rl_engine/kernels/ops/cuda/moe/mxfp8_act_quant.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from __future__ import annotations
+
+import pathlib
+import threading
+from typing import Any
+
+import torch
+from torch import Tensor
+
+import envs
```
Comment (KJLdefeated) [inline rl_engine/kernels/ops/cuda/moe/mxfp8_act_quant.py]: The current requirement package did not have this. Maybe adding a try catch on this path if the env package is not installed. We may add env package to required packages in the future.

**Review by KJLdefeated — CHANGES_REQUESTED** [review]
Well developed PR. Overall clean and clear. Just left some minor issues. Merge it after resolving the requests. Good Work!

File: rl_engine/kernels/ops/cuda/moe/mxfp8_act_quant.py (hunk 1 above)
Comment (ryankert01) [reply rl_engine/kernels/ops/cuda/moe/mxfp8_act_quant.py]: gotcha, solved

**Review by KJLdefeated — APPROVED** [review]
LGTM
